### PR TITLE
Mvc

### PR DESCRIPTION
**Pull request by walidovich**
CRUD mvc

### CODE REVIEW

File: src/test/java/dk/cngroup/trainings/spring/springassignment/controller/AnimalControllerTest.java [hunk 1]
```diff
@@ -37,6 +37,8 @@
 @RunWith(SpringRunner.class)
 public class AnimalControllerTest {
```
Comment (Wrent) [inline src/test/java/dk/cngroup/trainings/spring/springassignment/controller/AnimalControllerTest.java]: I know that this is just early stage, but even for the MVC controllers, there is a nice way in Spring how to test them. So don't forget to add tests.

You can still use the MockMvc, but instead of checking `jsonPath`. You can check what view is used, what is in the model, what is not in the model, etc. Some examples: https://www.petrikainulainen.net/programming/spring-framework/unit-testing-of-spring-mvc-controllers-normal-controllers/

File: src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebController.java [hunk 2]
```diff
@@ -1,19 +1,17 @@
...
 
-import javax.validation.Valid;
+import java.util.List;
 
 @Controller
-@RequestMapping("/web/animals")
+@RequestMapping("/web/animal")
 public class AnimalWebController {
 
 	private AnimalService animalService;
```
Comment (Wrent) [inline src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebController.java]: `animalService` can be `final` here. It is a good practise to have dependencies as final, so nobody can switch them once a class is instantiated (applies to other parts of app too)

File: src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebController.java [hunk 3]
```diff
@@ -22,32 +20,46 @@ public AnimalWebController(AnimalService animalService) {
...
+		ModelAndView modelAndView = new ModelAndView("animal/animal_form");
+		Animal animal = animalService.getAnimalById(id);
+		modelAndView.addObject("animal", animal);
 		return modelAndView;
 	}
+
+	@PostMapping("/save")
+	public ModelAndView save(@ModelAttribute("animal") Animal animal)
+			throws AnimalNotFoundException, InvalidAnimalException {
+		if (animal != null && animal.getId() != null) {
```
Comment (Wrent) [inline src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebController.java]: if `animal == null` then you still call `addAnimal`, it is probably not correct, right?

File: src/main/resources/templates/careTaker/careTaker_form.html [hunk 4]
```diff
@@ -6,7 +6,8 @@
 </head>
 <body>
 <h2>Animal Form:</h2>
```
Comment (Wrent) [inline src/main/resources/templates/careTaker/careTaker_form.html]: this is not Animal form, right? :D

**Review by Wrent — COMMENTED** [review]
Actually it looks really good, nice and simple :).

You can try working on tests as I wrote in a comment.

Also there is no error handling, the application throws some 500 error, if I dont write any name to an animal, it would be nice to have some error message directly in the form :)

File: src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebController.java (hunk 2 above)
Comment (walidovich) [reply src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebController.java]: Nice touch.

File: src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebController.java (hunk 3 above)
Comment (walidovich) [reply src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebController.java]: /add and /update both will lead to save, and save then decides if we are adding a new entity or updating it upon the animal id field.
The "if (animal!=null)" was just an extra, and useless, check. So I removed it.
I'm interested if you have a more proper way to handle the mvc crud operations.

File: src/main/resources/templates/careTaker/careTaker_form.html (hunk 4 above)
Comment (walidovich) [reply src/main/resources/templates/careTaker/careTaker_form.html]: Copy/Paste fail hhhhh

File: src/main/java/dk/cngroup/trainings/spring/springassignment/controller/advice/RESTControllerException.java [hunk 5]
```diff
@@ -3,11 +3,11 @@
 import dk.cngroup.trainings.spring.springassignment.exception.*;
 import org.springframework.http.HttpStatus;
 import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Controller;
 import org.springframework.web.bind.annotation.ControllerAdvice;
 import org.springframework.web.bind.annotation.ExceptionHandler;
-import org.springframework.web.bind.annotation.RestController;
 
-@ControllerAdvice(annotations = RestController.class)
+@ControllerAdvice(annotations = Controller.class)
```
Comment (Wrent) [inline src/main/java/dk/cngroup/trainings/spring/springassignment/controller/advice/RESTControllerException.java]: Ok, now you have some kind of error handling, but would not it be better if you actually got the error message where it happened? E.g. when you add new animal and the name is too short, it should show the form again (with the values you sent) and with some error message above or below the form. Don't you think?

File: src/test/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+				.andExpect(status().isOk())
+				.andExpect(model().attribute("animal", hasProperty("id", nullValue())))
+				.andExpect(view().name(viewPackage + "animal_form"));
+
+		verifyZeroInteractions(animalService);
+	}
+	// There is no Sad path for requesting the "/add"
+
+	@Test
+	public void saveNewAnimalTest() throws Exception {
```
Comment (Wrent) [inline src/test/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebControllerTest.java]: what about Sad path here? But I guess that based on my previous comment, the error handling will change a bit, so you can consider that (or you could try TDD approach for that :D)

File: src/main/java/dk/cngroup/trainings/spring/springassignment/controller/advice/RESTControllerException.java (hunk 5 above)
Comment (walidovich) [reply src/main/java/dk/cngroup/trainings/spring/springassignment/controller/advice/RESTControllerException.java]: Yeah, I've that in mind, I just wanted to implement the basic functionalities then I will head for error handling. An Yes you're right. error of fields validation should return to the same form.

Comment (walidovich) [reply src/main/java/dk/cngroup/trainings/spring/springassignment/controller/advice/RESTControllerException.java]: Yeah, I've that in mind. I wanted to finish all basic functionalities first then head to error handling.
You're right, error in fields validation should return to the same form.

File: src/test/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebControllerTest.java (hunk 6 above)
Comment (walidovich) [reply src/test/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebControllerTest.java]: Requesting that path only redirect to the form and doesn't interact with the service. So there is no sad path fortunately.

Comment (Wrent) [reply src/test/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebControllerTest.java]: well then you can test that it will do the redirect, right? But as I said, that will probably change when you do the error handling inside the form

File: src/test/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebControllerTest.java [hunk 7]
```diff
@@ -52,10 +52,10 @@ public void setUp() {
 
 	@Test
 	public void getAnimalsTest() throws Exception {
-		// Happy path
+		// Happy URL_PATH
```
Comment (Wrent) [inline src/test/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebControllerTest.java]: ahh, refactoring in IDEA with commnents included :)

File: src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebController.java [hunk 8]
```diff
@@ -1,53 +1,150 @@
...
 
-	@PostMapping("")
-	public ModelAndView addAnimal(@ModelAttribute @Valid Animal animal) {
-		ModelAndView modelAndView = new ModelAndView();
-		try {
+	@GetMapping("/add")
+	public ModelAndView add() {
+		ModelAndView modelAndView = new ModelAndView(VIEW_PATH + "/animal_form");
+		Animal animal = new Animal();
+		animal.setId(0L);
```
Comment (Wrent) [inline src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebController.java]: why is this here? it probably should not be here (maybe move it to Animal construtor?

File: src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebController.java [hunk 9]
```diff
@@ -1,53 +1,150 @@
...
+		return new ModelAndView("redirect:/web/animal/list/" + animalId);
+	}
+
+	@GetMapping("/list/{id}")
+	public ModelAndView getAnimalById(@PathVariable("id") Long id) throws AnimalNotFoundException {
+		Animal animal = animalService.getAnimalById(id);
+		ModelAndView modelAndView = new ModelAndView(VIEW_PATH + "/animal_details");
+		modelAndView.addObject("animal", animal);
+		CareTaker careTaker = new CareTaker();
+		careTaker.setId(0L);
```
Comment (Wrent) [inline src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebController.java]: same as above

File: src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/CareTakerWebController.java [hunk 10]
```diff
@@ -1,53 +1,149 @@
...
 
-	@PostMapping("")
-	public ModelAndView addCareTaker(@ModelAttribute @Valid CareTaker careTaker) {
-		ModelAndView modelAndView = new ModelAndView();
-		try {
+	@GetMapping("/add")
+	public ModelAndView add() {
+		ModelAndView modelAndView = new ModelAndView(VIEW_PATH + "/careTaker_form");
+		CareTaker careTaker = new CareTaker();
+		careTaker.setId(0L);
```
Comment (Wrent) [inline src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/CareTakerWebController.java]: same as above

File: src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/CareTakerWebController.java [hunk 11]
```diff
@@ -1,53 +1,149 @@
...
+		return new ModelAndView("redirect:/web/careTaker/list/" + careTakerId);
+	}
+
+	@GetMapping("/list/{id}")
+	public ModelAndView getCareTakerById(@PathVariable("id") Long id) throws CareTakerNotFoundException {
+		ModelAndView modelAndView = new ModelAndView(VIEW_PATH + "/careTaker_details");
+		CareTaker careTaker = careTakerService.getCareTakerById(id);
+		modelAndView.addObject("careTaker", careTaker);
+		Animal animal = new Animal();
+		animal.setId(0L);
```
Comment (Wrent) [inline src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/CareTakerWebController.java]: same as above

File: src/main/resources/templates/layouts/animal_details_layout.html [hunk 12]
```diff
@@ -0,0 +1,27 @@
+<!DOCTYPE html>
+<html xmlns:layout="http://www.w3.org/1999/xhtml" xmlns:th="http://www.thymeleaf.org">
```
Comment (Wrent) [inline src/main/resources/templates/layouts/animal_details_layout.html:2]: Format this file please, so it is better readable

File: src/main/resources/templates/layouts/careTaker_details_layout.html [hunk 13]
```diff
@@ -0,0 +1,27 @@
+<!DOCTYPE html>
```
Comment (Wrent) [inline src/main/resources/templates/layouts/careTaker_details_layout.html:1]: Format

File: src/main/resources/templates/layouts/header_up_footer_down_layout.html [hunk 14]
```diff
@@ -0,0 +1,20 @@
+<!DOCTYPE html>
```
Comment (Wrent) [inline src/main/resources/templates/layouts/header_up_footer_down_layout.html:1]: Format

**Review by Wrent — APPROVED** [review]
Nice, I think it look good, I have just some minor comments so fix them and merge this :)

File: src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebController.java (hunk 8 above)
Comment (walidovich) [reply src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebController.java]: Yes I will.

File: src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebController.java (hunk 9 above)
Comment (walidovich) [reply src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/AnimalWebController.java]: Yes I will.

File: src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/CareTakerWebController.java (hunk 11 above)
Comment (walidovich) [reply src/main/java/dk/cngroup/trainings/spring/springassignment/controller/web/CareTakerWebController.java]: Sir, yes Sir.

File: src/main/resources/templates/layouts/animal_details_layout.html (hunk 12 above)
Comment (walidovich) [reply src/main/resources/templates/layouts/animal_details_layout.html:2]: I will. If you mean putting each name space in different line.
